Migrate `:libs:processors` to `ksp`

wzieba opened the pull request:
## Description 
This PR rewrites `:libs:processors` annotation processor from `kapt` to `ksp`, to unblock `ksp` migration in the whole project.

Please mind that this PR goes into feature branch `ksp`, not to `trunk` yet.

## To Test:

<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

### TC1: Validate if generated classes did not change
- [x] Passed

1. Check out ec1baba114bba7d8eb6557fff1f48c581ab8870b
2. Run `clean WordPress:kaptWordpressWasabiDebugKotlin`
3. Save the content of generated classes somewhere. Those classes are
      - `RemoteFeatureConfigCheck`
      - `FeaturesInDevelopment`
      - `RemoteFeatureConfigDefaults`
      - `RemoteFieldConfigDefaults`
4. Check out this branch 
5. Run `clean WordPress:kspWordpressWasabiDebugKotlin`
6. Look for the same classes as above (they'll be in a different directory). **Assert that the content of those classes did not change** Please mind, that, unfortunately, the order of generated code is different. Hint: you can sort the files e.g. alphabetically to make it easier to compare.

### TC2: Incremental processing: add and remove a new development feature
- [x] Passed
1. Add a new remote feature class, in any place in the codebase, e.g. 
```kotlin
 @FeatureInDevelopment
 class Foobar
```
2. Run `WordPress:kspWordpressWasabiDebugKotlin`
3. **Assert `Foobar` is present in `FeaturesInDevelopment` generated file**
4. Remove some `@FeatureInDevelopment`, e.g. `LandOnTheEditorFeatureConfig`
5. Run `WordPress:kspWordpressWasabiDebugKotlin`
6. **Assert this feature is not present in `FeaturesInDevelopment` generated file**


### Other cases
In TC1 we validate content. In TC2 we validate incremental processing logic. Output files for other than `@FeaturesInDevelopment` annotations have the same setting - aggregating, so there's no strong need to test it.

But if you want to make sure, you can repeat TC2 with other annotations: `@Experiment`, `@Feature` and `@RemoteFieldDefaultGenerater`.

-----

## Regression Notes

1. Potential unintended areas of impact
    - Code generation of classes:
      - `RemoteFeatureConfigCheck`
      - `FeaturesInDevelopment`
      - `RemoteFeatureConfigDefaults`
      - `RemoteFieldConfigDefaults`

8. What I did to test those areas of impact (or what existing automated tests I relied on)

    - I wrote unit tests and smoke tested the app (instructions below)

9. What automated tests I added (or what prevented me from doing so)

    - Unit tests:
      - #20213 
      - #20227

-----

## PR Submission Checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist:

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

wzieba committed:
build: bump kotlin to 1.9.22, introduce ksp in libs:processors

wzieba committed:
refactor: rewrite `RemoteConfigProcessor` to ksp

dangermattic commented:
<!--
  1 Error: Please add tests for class <co...
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>RemoteConfigProcessorProvider</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
feat: specify correct `originatingFiles` to optimize ksp output invalidation

wzieba committed:
fix: do not generate classes if inputs are empty

This commit brings back behaviour from previous, kapt implementation, of not generating classes if inputs for them are empty.

This solves the problem of generating classes with empty content in different sourceset, e.g. `androidTest`. You can reproduce such problem in 80a80aa61c01a2c8a1e84fccd9a3ead307be8b7f

wzieba committed:
Merge branch 'kapt_processing_tests' into migrate_processors_to_ksp

wzieba committed:
Merge branch 'kapt_processing_tests' into migrate_processors_to_ksp

wzieba committed:
test: adjust test setup for ksp usage

wzieba committed:
Merge branch 'ksp' into migrate_processors_to_ksp

wzieba commented on libs/processors/src/main/java/org/wordpress/android/processor/RemoteConfigProcessor.kt:
```diff
@@ -1,128 +1,128 @@
+class RemoteConfigProcessor(
+    private val codeGenerator: CodeGenerator,
+) : SymbolProcessor {
+    /**
+     * See: https://github.com/google/ksp/issues/797#issuecomment-1041127747
+     * Also: https://github.com/google/ksp/blob/a0cd7774a7f65cec45a50ecc8960ef5e4d47fc21/examples/playground/test-processor/src/main/kotlin/TestProcessor.kt#L20
+     */
+    private var invoked = false
```
In the case of this processor, we only one need round - generated files do not depend on each other or any other processors.

mkevins replied:
Thanks for the explanation. Minor suggestion: we could add your GitHub comment as a code comment (above the links to the discussion and example source), in case this assumption is later invalidated.

wzieba replied:
Sure! Added in 0292847e39d9e704b117c6f3c53fb494881ad6a1

wzieba commented on libs/processors/src/main/java/org/wordpress/android/processor/RemoteConfigProcessor.kt:
```diff
@@ -1,128 +1,128 @@
+            RemoteFeatureConfigDefaultsBuilder(defaults).getContent()
+                .writeTo(
+                    codeGenerator,
+                    aggregating = true,
```
All output files are set to `aggregating` strategy for incremental processing. They have to, as we gather annotations from all available files in the codebase. 

See https://kotlinlang.org/docs/ksp-incremental.html#aggregating-vs-isolating for more details.

wzieba committed:
tests: simplify RemoteConfigProcessorTest

Remove the, now not necessairy, feature in every test. It's unnecessary now as condition was updated in RemoteConfigProcessor

mkevins commented on libs/processors/src/test/kotlin/org/wordpress/android/processor/RemoteConfigProcessorTest.kt:
```diff
@@ -32,12 +34,7 @@ class RemoteConfigProcessorTest {
-        val result = compile(
-            listOf(
-                remoteFieldA,
-                featureA, /* adding a feature, as without it, annotation processor won't start */
```
Nice that we are able to fix this 😄

mkevins reviewed (approved):
Thank you for migrating to KSP! The code changes look good, and I tested the scenarios described, and everything is working as expected. I did noticed, as you already mentioned that the generated outputs have a different order, but that should not affect anything.

I left one minor suggestion about adding more details to a code comment, in case multiple passes are required at some point in the future.

wzieba committed:
docs: add single-round processor comment

wzieba committed:
style: fix checkstyle
